File: change-risk-score-py/score.py
#!/usr/bin/env python3
"""Compute CRAP change-risk scores per function for a Python project.

CRAP(m) = comp(m)**2 * (1 - cov(m)/100)**3 + comp(m)

Complexity comes from `radon cc -j`. Coverage is read from a Cobertura XML
file produced by coverage.py. Per-function coverage is computed by intersecting
the function's line range (from radon) with the per-line hit counts in the
coverage XML — this avoids depending on the optional <methods> element, which
coverage.py does not always emit.
"""

import argparse
import json
import os
import subprocess
import sys
import xml.etree.ElementTree as ET
from pathlib import Path


def run_radon(source_path: str) -> dict:
    result = subprocess.run(
        ["radon", "cc", "-j", "-s", source_path],
        capture_output=True,
        text=True,
        check=True,
    )
    return json.loads(result.stdout) if result.stdout.strip() else {}


def parse_coverage(coverage_xml: str) -> dict[str, dict[int, int]]:
    tree = ET.parse(coverage_xml)
    root = tree.getroot()
    files: dict[str, dict[int, int]] = {}
    for cls in root.iter("class"):
        filename = cls.get("filename")
        if not filename:
            continue
        lines: dict[int, int] = {}
        for line in cls.findall("lines/line"):
            number = line.get("number")
            hits = line.get("hits")
            if number is None or hits is None:
                continue
            lines[int(number)] = int(hits)
        files[norm_path(filename)] = lines
    return files


def norm_path(path: str) -> str:
    return Path(path).as_posix()


def find_file(coverage: dict[str, dict[int, int]], filename: str):
    key = norm_path(filename)
    if key in coverage:
        return coverage[key]
    # Suffix-match fallback: handles cases where one side is rooted at the
    # repo and the other at a sub-package.
    for candidate, lines in coverage.items():
        if candidate.endswith(key) or key.endswith(candidate):
            return lines
    return None


def function_coverage(lines_map, start: int, end: int):
    if lines_map is None:
        return None
    relevant = [hits for n, hits in lines_map.items() if start <= n <= end]
    if not relevant:
        return None
    covered = sum(1 for h in relevant if h > 0)
    return (covered / len(relevant)) * 100.0


def crap(complexity: int, coverage_pct: float) -> float:
    return complexity * complexity * (1 - coverage_pct / 100.0) ** 3 + complexity


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--source-path", required=True)
    parser.add_argument("--coverage-xml", required=True)
    parser.add_argument("--threshold", type=float, required=True)
    parser.add_argument("--top", default="", help="Limit PR comment table to N worst rows")
    parser.add_argument("--missing-policy", choices=["pessimistic", "optimistic", "skip"], default="pessimistic")
    parser.add_argument("--output-file", required=True)
    parser.add_argument("--violations-file", required=True)
    args = parser.parse_args()

    if not Path(args.coverage_xml).exists():
        print(f"error: coverage XML not found at {args.coverage_xml}", file=sys.stderr)
        return 2

    radon_data = run_radon(args.source_path)
    coverage_data = parse_coverage(args.coverage_xml)

    scores = []
    for filename, items in radon_data.items():
        if isinstance(items, dict) and "error" in items:
            continue
        if not isinstance(items, list):
            continue
        lines_map = find_file(coverage_data, filename)
        for item in items:
            if item.get("type") not in ("function", "method"):
                continue
            name = item["name"]
            classname = item.get("classname")
            full_name = f"{classname}.{name}" if classname else name
            complexity = int(item["complexity"])
            start = int(item["lineno"])
            end = int(item.get("endline", start))

            coverage_pct = function_coverage(lines_map, start, end)
            if coverage_pct is None:
                if args.missing_policy == "skip":
                    continue
                coverage_pct = 0.0 if args.missing_policy == "pessimistic" else 100.0

            scores.append({
                "file": norm_path(filename),
                "function": full_name,
                "line": start,
                "complexity": complexity,
                "coverage": coverage_pct,
                "crap": crap(complexity, coverage_pct),
            })

    scores.sort(key=lambda s: s["crap"], reverse=True)
    violations = [s for s in scores if s["crap"] > args.threshold]

    top_n = int(args.top) if args.top.strip() else len(scores)
    displayed = scores[:top_n]

    lines = []
    lines.append("## Change Risk Score (Python)")
    lines.append("")
    lines.append(
        f"Threshold **{args.threshold:g}** · Functions analyzed **{len(scores)}** · "
        f"Above threshold **{len(violations)}**"
    )
    lines.append("")

    if displayed:
        lines.append("| CRAP | Complexity | Coverage | Function | Location |")
        lines.append("|---:|---:|---:|---|---|")
        for s in displayed:
            mark = " :warning:" if s["crap"] > args.threshold else ""
            lines.append(
                f"| {s['crap']:.1f}{mark} | {s['complexity']} | {s['coverage']:.1f}% | "
                f"`{s['function']}` | `{s['file']}:{s['line']}` |"
            )
        if len(scores) > len(displayed):
            lines.append("")
            lines.append(f"_…and {len(scores) - len(displayed)} more functions._")
    else:
        lines.append("_No functions analyzed._")

    Path(args.output_file).write_text("\n".join(lines) + "\n")

    with open(args.violations_file, "w") as f:
        for v in violations:
            f.write(
                f"{v['file']}:{v['line']} {v['function']} "
                f"CRAP={v['crap']:.1f} complexity={v['complexity']} coverage={v['coverage']:.1f}%\n"
            )

    print(f"Analyzed {len(scores)} functions; {len(violations)} above threshold {args.threshold:g}.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: change-risk-score-js/score.mjs
#!/usr/bin/env node
// Compute CRAP change-risk scores per function for a JS/TS (and optionally
// Svelte) project. Complexity comes from ESLint's built-in `complexity` rule
// using the Node API. Coverage is read from an lcov.info file (the format
// emitted by c8, Vitest, Jest, nyc). Per-function coverage is computed by
// intersecting each function's line range with lcov's DA records.

import { ESLint } from 'eslint';
import { readFile, writeFile } from 'node:fs/promises';
import { relative } from 'node:path';

function parseArgs(argv) {
  const args = {};
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (!a.startsWith('--')) continue;
    const key = a.slice(2);
    const next = argv[i + 1];
    if (next !== undefined && !next.startsWith('--')) {
      args[key] = next;
      i++;
    } else {
      args[key] = 'true';
    }
  }
  return args;
}

function normPath(p) {
  return p.replace(/\\/g, '/').replace(/^\.\//, '');
}

function parseLcov(content) {
  const files = new Map();
  let current = null;
  for (const line of content.split(/\r?\n/)) {
    if (line.startsWith('SF:')) {
      current = { filename: line.slice(3), lines: new Map() };
    } else if (line.startsWith('DA:') && current) {
      const [n, h] = line.slice(3).split(',');
      current.lines.set(Number(n), Number(h));
    } else if (line === 'end_of_record' && current) {
      files.set(normPath(current.filename), current.lines);
      current = null;
    }
  }
  return files;
}

function findFile(coverage, filename) {
  const key = normPath(filename);
  if (coverage.has(key)) return coverage.get(key);
  for (const [candidate, lines] of coverage) {
    if (candidate.endsWith(key) || key.endsWith(candidate)) return lines;
  }
  return null;
}

function functionCoverage(linesMap, start, end) {
  if (!linesMap) return null;
  let total = 0;
  let covered = 0;
  for (const [n, h] of linesMap) {
    if (n >= start && n <= end) {
      total++;
      if (h > 0) covered++;
    }
  }
  if (total === 0) return null;
  return (covered / total) * 100;
}

function crap(complexity, coveragePct) {
  return complexity * complexity * Math.pow(1 - coveragePct / 100, 3) + complexity;
}

const FN_NAME_RE = /(?:Async generator function|Async generator method|Generator function|Generator method|Async arrow function|Arrow function|Async function|Async method|Function|Method)\s*(?:'([^']+)')?/;
const COMPLEXITY_RE = /complexity of (\d+)/;

function extractFnInfo(message) {
  const nameMatch = message.match(FN_NAME_RE);
  const complexityMatch = message.match(COMPLEXITY_RE);
  return {
    name: (nameMatch && nameMatch[1]) || '<anonymous>',
    complexity: complexityMatch ? Number(complexityMatch[1]) : 0,
  };
}

async function buildConfig({ svelte }) {
  const tsParserMod = await import('@typescript-eslint/parser');
  const tsParser = tsParserMod.default ?? tsParserMod;

  const config = [
    {
      files: ['**/*.js', '**/*.jsx', '**/*.mjs', '**/*.cjs'],
      languageOptions: { ecmaVersion: 'latest', sourceType: 'module' },
      rules: { complexity: ['error', 0] },
    },
    {
      files: ['**/*.ts', '**/*.tsx', '**/*.mts', '**/*.cts'],
      languageOptions: { parser: tsParser, ecmaVersion: 'latest', sourceType: 'module' },
      rules: { complexity: ['error', 0] },
    },
  ];

  if (svelte) {
    const svelteParserMod = await import('svelte-eslint-parser');
    const svelteParser = svelteParserMod.default ?? svelteParserMod;
    config.push({
      files: ['**/*.svelte'],
      languageOptions: {
        parser: svelteParser,
        parserOptions: { parser: tsParser },
      },
      rules: { complexity: ['error', 0] },
    });
  }

  return config;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const sourcePath = args['source-path'];
  const coveragePath = args['coverage-lcov'];
  const threshold = Number(args['threshold']);
  const top = args['top'] && args['top'].trim() !== '' ? Number(args['top']) : Infinity;
  const missingPolicy = args['missing-policy'] || 'pessimistic';
  const outputFile = args['output-file'];
  const violationsFile = args['violations-file'];
  const svelte = args['svelte'] === 'true';
  const extensions = (args['extensions'] || 'js,jsx,mjs,cjs,ts,tsx,mts,cts')
    .split(',')
    .map((e) => '.' + e.trim())
    .filter((e) => e.length > 1);
  if (svelte && !extensions.includes('.svelte')) extensions.push('.svelte');

  let coverageText;
  try {
    coverageText = await readFile(coveragePath, 'utf8');
  } catch (e) {
    console.error(`error: could not read lcov coverage file at ${coveragePath}: ${e.message}`);
    process.exit(2);
  }
  const coverageData = parseLcov(coverageText);

  const overrideConfig = await buildConfig({ svelte });
  const eslint = new ESLint({
    overrideConfigFile: true,
    overrideConfig,
    errorOnUnmatchedPattern: false,
  });

  const results = await eslint.lintFiles([sourcePath]);

  const scores = [];
  for (const result of results) {
    const fileRel = normPath(relative(process.cwd(), result.filePath));
    if (!extensions.some((ext) => fileRel.endsWith(ext))) continue;
    const linesMap = findFile(coverageData, fileRel);
    for (const m of result.messages) {
      if (m.ruleId !== 'complexity') continue;
      const { name, complexity } = extractFnInfo(m.message);
      const start = m.line;
      const end = m.endLine ?? start;
      let coveragePct = functionCoverage(linesMap, start, end);
      if (coveragePct == null) {
        if (missingPolicy === 'skip') continue;
        coveragePct = missingPolicy === 'pessimistic' ? 0 : 100;
      }
      scores.push({
        file: fileRel,
        function: name,
        line: start,
        complexity,
        coverage: coveragePct,
        crap: crap(complexity, coveragePct),
      });
    }
  }

  scores.sort((a, b) => b.crap - a.crap);
  const violations = scores.filter((s) => s.crap > threshold);
  const displayed = scores.slice(0, Math.min(top, scores.length));

  const lines = [];
  lines.push('## Change Risk Score (JavaScript/TypeScript)');
  lines.push('');
  lines.push(
    `Threshold **${threshold}** · Functions analyzed **${scores.length}** · Above threshold **${violations.length}**`,
  );
  lines.push('');
  if (displayed.length) {
    lines.push('| CRAP | Complexity | Coverage | Function | Location |');
    lines.push('|---:|---:|---:|---|---|');
    for (const s of displayed) {
      const mark = s.crap > threshold ? ' :warning:' : '';
      lines.push(
        `| ${s.crap.toFixed(1)}${mark} | ${s.complexity} | ${s.coverage.toFixed(1)}% | \`${s.function}\` | \`${s.file}:${s.line}\` |`,
      );
    }
    if (scores.length > displayed.length) {
      lines.push('');
      lines.push(`_…and ${scores.length - displayed.length} more functions._`);
    }
  } else {
    lines.push('_No functions analyzed._');
  }

  await writeFile(outputFile, lines.join('\n') + '\n');
  await writeFile(
    violationsFile,
    violations.length
      ? violations
          .map(
            (v) =>
              `${v.file}:${v.line} ${v.function} CRAP=${v.crap.toFixed(1)} complexity=${v.complexity} coverage=${v.coverage.toFixed(1)}%`,
          )
          .join('\n') + '\n'
      : '',
  );

  console.log(`Analyzed ${scores.length} functions; ${violations.length} above threshold ${threshold}.`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
